File: common/src/main/java/com/ssblur/scriptor/effect/PhasingStatusEffect.java
package com.ssblur.scriptor.effect;

import com.ssblur.scriptor.blockentity.PhasedBlockBlockEntity;
import net.minecraft.core.Vec3i;
import net.minecraft.world.effect.MobEffect;
import net.minecraft.world.effect.MobEffectCategory;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;

public class PhasingStatusEffect extends MobEffect {
  public PhasingStatusEffect() {
    super(MobEffectCategory.NEUTRAL, 0x2d0096);
  }

  public PhasingStatusEffect(MobEffectCategory mobEffectCategory, int i) {
    super(mobEffectCategory, i);
  }

  @Override
  public boolean applyEffectTick(LivingEntity entity, int amplifier) {
    if(entity instanceof Player player)
      phase(player, 0);
    return true;
  }

  public static void phase(Player entity, int bottom) {
    for (int x = 1; x >= -1; x--)
      for (int y = (int) Math.ceil(entity.getEyeHeight()); y >= bottom; y--)
        for (int z = 1; z >= -1; z--)
          PhasedBlockBlockEntity.phase(entity.level(), entity.blockPosition().offset(new Vec3i(x, y, z)));
  }

  @Override
  public boolean shouldApplyEffectTickThisTick(int i, int j) {
    return true;
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/effect/PhasingStatusEffect.kt
package com.ssblur.scriptor.effect

import com.ssblur.scriptor.blockentity.PhasedBlockBlockEntity
import net.minecraft.core.Vec3i
import net.minecraft.world.effect.MobEffect
import net.minecraft.world.effect.MobEffectCategory
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import kotlin.math.ceil

open class PhasingStatusEffect: MobEffect {
  constructor(): super(MobEffectCategory.NEUTRAL, 0x2d0096)

  constructor(mobEffectCategory: MobEffectCategory, i: Int): super(mobEffectCategory, i)

  override fun applyEffectTick(entity: LivingEntity, amplifier: Int): Boolean {
    if (entity is Player) phase(entity, 0)
    return true
  }

  override fun shouldApplyEffectTickThisTick(i: Int, j: Int): Boolean {
    return true
  }

  companion object {
    fun phase(entity: Player, bottom: Int) {
      for (x in 1 downTo -1) for (y in ceil(entity.eyeHeight.toDouble()).toInt() downTo bottom) for (z in 1 downTo -1) PhasedBlockBlockEntity.phase(
        entity.level(), entity.blockPosition().offset(
          Vec3i(x, y, z)
        )
      )
    }
  }
}